### ISSUE TITLE
Defect: change team for subset of images eventually deadlocks

### ISSUE THREAD

**Issue by nathanweeks** [sourceryinstitute/OpenCoarrays#610]
# Defect/Bug Report #

In the following test, images corresponding to only one of two teams resulting from a FORM TEAM statement executes a subsequent CHANGE TEAM statement. There is a deadlock in the mpi_caf.c (in change_team()) due to an MPI_Barrier involving the communicator associated with the parent team.

```Fortran
program teams_subset
  use iso_fortran_env, only : team_type
  implicit none

  type(team_type) :: team
  integer :: initial_team_image, max_image(2), min_image(2), myteam

  initial_team_image = this_image()

  if (initial_team_image == 1 .or. initial_team_image == num_images()) then
    myteam = 1
  else
    myteam = 2
  end if

  form team (myteam, team)

  if (myteam == 1) then
    change team(team)
      max_image = [initial_team_image, this_image()]
      call co_max(max_image)
      min_image = [initial_team_image, this_image()]
      call co_min(min_image)
    end team
    if (any(min_image /= [1, 1]) .or. any(max_image /= [num_images(), 2])) then
      write(*,*) "Test failed."
      error stop
    end if
  end if

  sync all

  if (initial_team_image == 1) write(*,*) "Test passed."

end program
```

- OpenCoarrays Version: 2.3.1
- Fortran Compiler: gfortran 8.2.0
- C compiler used for building lib: gcc 8.2.0
- Installation method: MacPorts
- Output of `uname -a`: Darwin my-mac.local 15.6.0 Darwin Kernel Version 15.6.0: Thu Jun 21 20:07:40 PDT 2018; root:xnu-3248.73.11~1/RELEASE_X86_64 x86_64
- MPI library being used: MPICH v3.3rc1
- Machine architecture and number of physical cores: x86, 2
- Version of CMake: 3.12.4

## Observed Behavior ##

When OpenCoarrays is configured with `-DCMAKE_BUILD_TYPE=Debug`:

```
$ caf teams_subset.f90
$ cafrun -np 3 ./a.out
2/3: _gfortran_caf_sync_all(1537) Entering sync all.
... hangs ...
```

## Expected Behavior ##

`Test passed.`

**Comment by nathanweeks** [comment]
@MichaelSiehl : is there wording in the Fortran 2018 standard that prohibits a CHANGE TEAM statement from being executed by only some of the images in the current team? In addition to the Note 2.16 mentioned in #611 , the only relevant wording I could find from a quick search of the N2146 draft seems to not prohibit this:

> 11.1.5.2 Execution of a CHANGE TEAM construct
...
> 5. Successful execution of a CHANGE TEAM statement performs an implicit synchronization of all images of the new team that is identified by team-value. All active images of the new team shall execute the same CHANGE TEAM statement. On each image of the new team, execution of the segment following the CHANGE TEAM statement is delayed until all other images of that team have executed the same statement the same number of times in the original team.

There appears to have been a requirement in an earlier TS that was removed: 

> Changes to features previously described by ISO/IEC TS 18508:2015:
> ...
> All images of the current team are no longer required to execute the same CHANGE TEAM statement.

In addition, _Modern Fortran Explained: Incorporating Fortran 2018_ explicitly supports the validity of a subset of teams executing a CHANGE TEAM construct on p. 384 (https://goo.gl/uQ1DdH):

 > Both the _change team_ and _end team_ statements are image control statements. The executing image and the other images of its new team synchronize at these statements. These images execute the same _change team_ statement. While we expect it to be usual for the images of the other teams to execute the construct, this is not required - they might continue to execute in the previously current team.

**Comment by MichaelSiehl** [comment]
@nathanweeks : The quote from _Modern Fortran explained_ brings it to the point: _"While we expect it to be usual for the images of the other teams to execute the construct, this is not required - they might continue to execute in the previously current team."_

OpenCoarrays does support this (see my above example). But that can't work with the initial team because it is the only team that does not have a previous team.
 
A previous version of the above quote (in the paper 'The new features of Fortran 2018' by John Reid) was: _"While we expect it to be usual for the images of the other **new** teams to execute the construct..."_
Maybe it was a mistake to delete the word **new** from this sentence, because the new teams are those that are formed with a FORM TEAM statement, before executing a CHANGE TEAM statement. This is exactly what your are doing with your example, and it is supported by OpenCoarays (except in the initial team). See the following example: I did modify my above example. This time I do form SubTeam_1 before the SELECT CASE statement with two teams on all images of SubTeam. Only in the CASE(1) block I do execute a CHANGE TEAM (SubTeam_1), while the CASE(2) block does  continue to execute in the previously current team (which is SubTeam):
```fortran
program Main
!
! you may compile and run with 15 coarray images
  !
  use,intrinsic :: ISO_FORTRAN_ENV, only: team_type
  implicit none
  !
  integer :: intTeamNumber
  type (team_type) :: MainTeam
  type (team_type) :: SubTeam
  type (team_type) :: SubTeam_1
!
!************************************************
! MainTeam **************************************
!************************************************
!
! split the 15 images of the initial team into 3 child teams consisting of 5 images each:
  if (this_image() <= 5) then
    intTeamNumber = 1
  else if (this_image() >= 6 .and. this_image() <= 10) then
    intTeamNumber = 2
  else if (this_image() >= 11) then
    intTeamNumber = 3
  end if
  !
form team (intTeamNumber, MainTeam) ! this creates the 3 child teams in MainTeam
!
change team (MainTeam)
!
!************************************************
! SubTeam ***************************************
!************************************************
!
! split the 5 images of the MainTeam's child teams into two distinct teams resp.:
  if (this_image() < 3) then
    intTeamNumber = 1
  else
    intTeamNumber = 2
  end if
!
form team (intTeamNumber, SubTeam) ! this generates 6 teams (2 new sub teams in the 3 main teams resp.)
change team (SubTeam)
!
! form SubTeam_1:
if (this_image() < 2) then
  intTeamNumber = 1
else
  intTeamNumber = 2
end if
form team (intTeamNumber, SubTeam_1)
!
SubTeam_select: select case (team_number())
!
case (1) SubTeam_select
  ! code for child team 1 of SubTeam:
  !**********************************************
  ! change to SubTeam_1 **************
  !**********************************************
  change team (SubTeam_1)
    write(*,*) 'image number: ', this_image(), ' of team number: ', team_number(), &
             ' in SubTeam_1 '
    !
  end team !(SubTeam_1)
  !***************
  !***************
  !***************
!*******************
case (2) SubTeam_select
  ! code for child team 2 of SubTeam:
  write(*,*) 'image number: ', this_image(), ' of team number: ', team_number(), &
             ' in SubTeam '
!
end select SubTeam_select
!
end team !(SubTeam)
!
!*************
!*************
!*************
!
end team !(MainTeam)
!
end program Main
```

**Comment by MichaelSiehl** [comment]
uuhh.. Your above quote was the mistake. The book does still contain the original version of the sentence. So the book is ok.
cheers

**Comment by rouson** [comment]
Does this issue need to be reopened?  I'm doing so just case. @nathanweeks and @MichaelSiehl one of you can close it if you agree that there is no remaining issue.  Do we need to revert PR #611?

**Comment by nathanweeks** [comment]
Apologies for the incorrect manual transcription of the relevant paragraph of Modern Fortran Explained (Google Books doesn't seem to allow copy & paste). I've added the missing "new" to the excerpt.

I certainly may be misinterpreting the quote or code, but I don't think the above example illustrates the scenario that the excerpt describes. The `form team (intTeamNumber, SubTeam_1)` statement creates two new teams; on TEAM_NUMBER() == 1 of the current team, all images of both new teams execute the `change team (SubTeam_1)` statement, rather than the images of only one of the two new teams. I would welcome any corrections to this interpretation, though.

PR #611 is somewhat distinct from the issue of the validity of the _teams_subset_ code (though its justification may be different depending on the outcome of this issue). OpenCoarrays 2.3.1 implemented CHANGE TEAM synchronization semantics per TS 18508:2015 (synchronizing all images in the team that calls CHANGE TEAM), rather than the (more performant) Fortran 2018 semantics as described in NOTE 11.6 (synchronizing images within each sibling team CHANGE TEAM).

**Comment by MichaelSiehl** [comment]
@rouson : This issue may require further discussion:
My above statement, _" But that can't work with the initial team because it is the only team that does not have a previous team."_, may be no valid reason to not allow for execution of a CHANGE TEAM statement on only a subset of the images of the initial team.

Instead, see the definition for the _initial team_ in section 3.145.2 of the Fortran 2018 standard (draft): 

> initial team: team existing at the beginning of program execution, consisting of all images

I think, if OpenCoarrays would allow for execution of CHANGE TEAM on only a subset of images of the initial team, it would conflict with this definition of _initial team_ . The images of the _initial team_ that do not execute a CHANGE TEAM statement would (theoretically) remain within the _initial team_, but at the same time the _initial team_, as it is defined by the standard, would cease to exist because it would no longer consist of all the images and program execution would be no longer at it's beginning (on the images that did execute the CHANGE TEAM statement). This is currently my main argument for rejecting execution of a CHANGE TEAM statement on only a subset of images of the _initial team_.

Best Regards

**Comment by MichaelSiehl** [comment]
@nathanweeks : 

> I certainly may be misinterpreting the quote or code, but I don't think the above example illustrates the scenario that the excerpt describes. The form team (intTeamNumber, SubTeam_1) statement creates two new teams; on TEAM_NUMBER() == 1 of the current team, all images of both new teams execute the change team (SubTeam_1) statement, rather than the images of only one of the two new teams. I would welcome any corrections to this interpretation, though.

We may have different interpretations of the Modern Fortran Explained quote _"While we expect it to be usual for the images of the other new teams to execute the construct, this is not required - they might continue to execute in the previously current team."_ And I believe to have identified the reason for this:

You want to do something like this (?):

! form SubTeam_1:
if (this_image() < 2) then
  intTeamNumber = 1
else
  intTeamNumber = 2
end if
form team (intTeamNumber, SubTeam_1)
!
  **if (intTeamNumber == 1) then**
  change team (SubTeam_1)
    write(*,*) 'image number: ', this_image(), ' of team number: ', team_number(), &
             ' in SubTeam_1 '
    !
  end team !(SubTeam_1)
  end if

The problem here is the if statement using the intTeamNumber integer, which leads to a synchronization failure of the CHANGE TEAM statement because the integer does not contain the current team number. Instead you should always use the intrinsic function team_number to get the current team number. With the FORM TEAM statement we do form new teams. But we do not change to the new teams and we do not change the team numbers. The current team, and the current team number, is still that of what you might expect to be the previous team already (in my example SubTeam). Therefore, the term **new team** refers only to the teams formed by the FORM TEAM statement, without having already changed to the new team and team number (this change is not before the execution of CHANGE TEAM). This is the basic idea behind my above code example and it works perfectly with OpenCoarrays. So, yes I do believe that OpenCoarray's current runtime behaviour does adhere with the Fortran 2018 standard. 

Best Regards

**Comment by MichaelSiehl** [comment]
Thus, my understanding of the quote is,
_"...the images of the other new teams..."_ are still the images of SubTeam (and not SubTeam_1) in my above example. 
And _"...they might continue to execute in the previously current team."_, here the previously current team is also SubTeam, this time after the CHANGE TEAM statement got executed on only a subset of the teams of SubTeam (the viewpoint is now from this subset of teams that are now in SubTeam_1).

**Comment by nathanweeks** [comment]
@MichaelSiehl:
 
> The images of the initial team that do not execute a CHANGE TEAM statement would (theoretically) rema   in within the initial team, but at the same time the initial team, as it is defined by the standard, would cease to exist because it would no longer consist of all the images and program execution would be  no longer at it's beginning (on the images that did execute the CHANGE TEAM statement). 

The initial team shouldn't cease to exist upon the execution a CHANGE TEAM construct; e.g., the definition for current team implies the initial team still exists after the completion of the first CHANGE TEAM construct (italics mine):

**3.145.1**
**current team**
team specified by the most recently executed CHANGE TEAM statement of a CHANGE TEAM construct that has not completed execution (11.1.5), or initial team if no CHANGE TEAM construct _is being executed_

> The problem here is the if statement using the intTeamNumber integer, which leads to a synchronization failure of the CHANGE TEAM statement because the integer does not contain the current team number. 

Per NOTE 11.6, _"Execution of a CHANGE TEAM statement includes a synchronization of the executing image with the other images that will be in the same team after execution of the CHANGE TEAM statement"_. Since all images on the current team with `intTeamNumber == 1` would be on the same team after execution of the CHANGE TEAM statement (and no other images), that synchronization can take place.

> Thus, my understanding of the quote is,
_"...the images of the other new teams..."_ are still the images of SubTeam (and not SubTeam_1) in my above example.

From the preceding several paragraphs, I *think* the text is referring to (in this case) new teams that are created from the FORM TEAM statement executed by all images in the current team (SubTeam); i.e., teams of SubTeam_1. But it would be beneficial to seek clarification from the author or other Fortran standards committee members that decided the semantics of teams; I could reach out to them if there are no other volunteers.

(and thanks for the discussion on these topics---they're important points to get right!)

**Comment by MichaelSiehl** [comment]
If you see the mistake, you should also be able to understand why you can't do, what you are intending to do, in the initial team: You need a set of teams, but the initial team comprises only a single team.

**Comment by nathanweeks** [comment]
@MichaelSiehl : in the original version of the code, the CHANGE TEAM construct should cause synchronization only of images on the same new team (in this case, images that executed `form team (myteam, team)` with `myteam == 1`), so it should not deadlock.

That's my interpretation for Fortran 2018, anyway. In TS 18508, CHANGE TEAM specified synchronization for all images in the current team, in which case there would have been a deadlock. 

A news entry on the WG5 web site (https://wg5-fortran.org/2017news.html) words the update to the CHANGE TEAM semantics in a manner that appears to support an interpretation of CHANGE TEAM as used in the original version of the code:

> **12 July 2017 
> The second CD ballot on the revision of the standard**
> ... 
> The most significant changes concerned the CHANGE TEAM construct. In the revised draft, there is no requirement for all the images of the current team to execute as members of new teams.

I couldn't find any wording in N2146 that suggests this doesn't apply when the current team is the initial team (though admittedly have not read through all > 600 pages with this topic in mind).

**Comment by MichaelSiehl** [comment]
@nathanweeks : The wording in the standards text, as well as that in Modern Fortran Explained, makes perfect sense if you try to see it from the perspective of my above code examples, and if you see it in an appropriate context. My above code examples do work with OpenCoarrays 2.2.0 and they already employ highly flexible use of Fortran 2018 teams.
The basic idea behind your code will not work, even if we do apply the changes of PR #611. To me, this is the main reason to open #614 to revert from the commits of #611.

Best Regards

**Comment by MichaelSiehl** [comment]
@rouson @nathanweeks I think this could be closed as well.

**Comment by zbeekman** [comment]
Closing this since it should have been resolved by #611. Please let me know if you think this needs to be reopened.
